Add S3FileSystem Write support

majetideepak opened the pull request:
S3WriteFile uses the Apache Arrow implementation as a reference.
AWS C++ SDK allows streaming writes via the MultiPart upload API.
Multipart upload allows you to upload a single object as a set of parts.
Each part is a contiguous portion of the object's data.
While AWS and Minio support different sizes for each
part (only requiring a minimum of 5MB), Certain object stores require that every
part be exactly equal (except for the last part). We set this to 10 MiB, so
that in combination with the maximum number of parts of 10,000, this gives a
file limit of 100k MiB (or about 98 GiB).
You can upload these object parts independently and in any order.
After all parts of your object are uploaded, Amazon S3 assembles these parts
and creates the object.
S3WriteFile is not thread-safe.
UploadPart is currently synchronous during append. Flush is no-op as append 
handles all the uploads.

Resolves: https://github.com/facebookincubator/velox/issues/4805

majetideepak commented:
CC: @akashsha1, @paul-amonson, @tigrux for early feedback.
I need to run some performance tests to wrap up this. I will share them.

paul-amonson commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+  void uploadPart(bool isLast = false) {
+    // Upload the part.
+    // Only the last part can be less than kMinPartSize.
+    VELOX_CHECK(isLast || (!isLast && (currentPartSize_ >= kMinPartSize)));
```
I see where removing the lines above in flush could cause this exception. What is the intended goal? Maybe flush() should be a NOOP since close() basically performs a flush() with uploadPart(true)?

majetideepak replied:
`flush()` is part of the WriteFile API and has to be implemented. https://github.com/facebookincubator/velox/blob/main/velox/common/file/File.h#L139
The idea is that `append()` keeps writing to the `currentPart_`, an in-memory buffer.
When the application invokes `flush()`, we upload a part, if the AWS S3 constraints satisfy `(i.e currentPartSize_ > 5 MiB)`. We then clear the `currentPart_` buffer and handle the new `append()`s as a new part.
`flush()` essentially allows the application to bound the in-memory write buffer.
Without it, we will end up buffering the whole file in memory.

paul-amonson replied:
Ahhhhh....ok I get it. Its a AWS constraint. Fair enough. :)

majetideepak replied:
I made a comment on AWS S3 constraints where  `kMinPartSize` is defined. I will probably move that on top as well.

akashsha1 commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+    if (currentPartSize_ < kMinPartSize) {
+      return;
+    }
+    uploadPart();
```
should this be within a loop, till all parts are flushed, or is that with the caller of the flush method?

majetideepak replied:
There is only a single part being populated via `append()` at a given time. When the caller invokes `flush`, the entire data buffered is uploaded as a single part. All the current and previous parts are on S3.

majetideepak commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+      auto objectMetadata = client_->HeadObject(request);
+      VELOX_CHECK(!objectMetadata.IsSuccess(), "File already exists");
```
Will Fix.

majetideepak commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+      /// which confuses some tools
+      /// (https://github.com/apache/arrow/issues/11934). So we instead default
+      /// to application/octet-stream which is less misleading.
+      request.SetContentType("application/octet-stream");
```
Will do.

majetideepak commented on velox/connectors/hive/storage_adapters/s3fs/tests/S3FileSystemTest.cpp:
```diff
@@ -344,3 +345,36 @@ TEST_F(S3FileSystemTest, logLevel) {
+  filesystems::S3FileSystem s3fs(hiveConfig);
+  s3fs.initializeClient();
+  auto writeFile = s3fs.openFileForWrite(s3File);
+  std::string dataContent =
```
Yes! there should be a lot more tests for various paths. I updated the description.
We can probably make `kMinPartSize`, `kMaxPartSize` configurable since different object stores have different limits.
Alibaba cloud says part must be `[100KB to 5GB]` https://www.alibabacloud.com/help/en/oss/user-guide/multipart-upload-12

majetideepak replied:
I added a test for the multi-part upload.

akashsha1 replied:
you can add a TODO for partsize configuration.

majetideepak commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+      auto result = outcome.GetResult();
+      Aws::S3::Model::CompletedPart part;
+
+      part.SetPartNumber(uploadState_.partNumber);
```
As of now, if any AWS SDK operation fails, a VELOX_ERROR is thrown. There is no retry. So this should be okay for now.
I will make a comment to implement retries in a follow-up PR. They are important.

tigrux commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -172,6 +179,184 @@ class S3ReadFile final : public ReadFile {
+ public:
+  explicit S3WriteFile(const std::string& path, Aws::S3::S3Client* client)
+      : client_(client) {
+    bucketAndKeyFromS3Path(path, bucket_, key_);
```
nit: Should it be named setBucketAndKeyFromS3Path?

majetideepak replied:
Fixed here https://github.com/facebookincubator/velox/pull/6528

majetideepak replied:
I named it `getBucketAndKeyFromS3Path` since the function is inferring those values.

tigrux reviewed (approved):
Very clean code.

paul-amonson reviewed (approved):
:)

akashsha1 commented on velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:
```diff
@@ -198,6 +207,214 @@ Aws::Utils::Logging::LogLevel inferS3LogLevel(std::string level) {
+    VELOX_CHECK(!closed_, "File is closed");
+    // 'flush' API should trigger uploadPart.
+    // But upload part if the maximum part size is reached.
+    if (currentPartSize_ + data.size() > kMaxPartSize) {
```
should this be >=

majetideepak replied:
`>=` should work according to the documentation. Will do that.

majetideepak commented:
@tigrux @akashsha1 @paul-amonson I did some tests at scale and had to fix the semantics a bit.
I observed that the query would fail after writing a 5GB file. The issue was the difference
 between 5 GiB (code limit) vs. GB (AWS limit) upper part limit. 
This made me realize that the current TableWriter uses only `append()` and never a `flush()` call.
So each part would have been 5GB big and was being buffered in memory before upload which is bad.
I looked at the latest Arrow S3FS implementation and saw that they use a fixed-size part to upload.
I changed the semantics to match this. This made more sense as the write memory is now fixed.
The multi-part upload API also does not match the `append()` + `flush()` APIs as it has the min and max part limitations.
Making the `append()` manage the uploads and `flush()` a no-op matches the multi-part upload semantics well.
I added some comments to reflect this change.

majetideepak commented:
With the new semantics, I verified that we can now write files larger than 5GiB.

majetideepak commented:
I also looked at the `gcs::ObjectWriteStream` documentation and it does something similar.
https://cloud.google.com/cpp/docs/reference/storage/latest/classgoogle_1_1cloud_1_1storage_1_1ObjectWriteStream

majetideepak commented:
@pedroerp, @kgpai can you please help merge this? Thanks.

majetideepak commented:
The linux-build failure is unrelated.
